File: 20230126_Thursday/20230126.rb
# Leetcode: 787. Cheapest Flights Within K Stops.
# https://leetcode.com/problems/cheapest-flights-within-k-stops/
# @param {Integer} n
# @param {Integer[][]} flights
# @param {Integer} src
# @param {Integer} dst
# @param {Integer} k
# @return {Integer}
# Error.
def find_cheapest_price(n, flights, src, dst, k)
  # Based on:
  # https://leetcode.com/problems/cheapest-flights-within-k-stops/discuss/686906/Python-Multipass-BFS-O(V2)-%2B-Dijkstra-with-SortedList-explained
  g = {}
  v = [[src,0,0]]
  mp = Float::INFINITY
  visited = {}
  visited[src] = true
  flights.each do |i,j,w|
    g[i] ||= []
    g[i].push([j,w])
  end
  while ! v.empty?
    c,vis,pr = v.shift()
    if pr <= mp && vis <= k && c != dst
      g[c].each do |j,np|
        next if visited[j]
        v.push([j,vis+1,pr + np])
      end
    end
    if dst == c
      mp = [mp,pr].min
    end
  end
  mp = -1 if Float::INFINITY == mp
  return mp
end

File: 20230126_Thursday/20230126.py
# Leetcode: 787. Cheapest Flights Within K Stops.
# https://leetcode.com/problems/cheapest-flights-within-k-stops/
class Solution:
    def findCheapestPrice(self, n, flights, src, dst, k):
        graph = defaultdict(list)
        dist = [[float("inf")] * (k+2) for _ in range(n)] 
        dist[src][0] = 0

        SList = SortedList([[0, 0, src]])
        
        for i, j, w in flights: 
            graph[i].append([j, w])
        
        while SList:
            price, visited, city = SList.pop(0)

            if city == dst: return price
            
            if visited <= k:
                for neibh, price_neibh in graph[city]: 
                    candidate = dist[city][visited] + price_neibh
                    if candidate <= dist[neibh][visited + 1]:
                        dist[neibh][visited + 1] = candidate
                        SList.add([price + price_neibh, visited + 1, neibh])
                        
        return -1
